spec/lib/workers/heartbeat_spec.rb

```diff
@@ -69,4 +69,25 @@
       it_should_behave_like "heartbeat file checker", other_heartbeat_file
     end
   end
+
+  describe "#post_message_for_workers" do
+    let(:started_worker)  { FactoryBot.create(:miq_ui_worker, :status => MiqWorker::STATUS_STARTED,  :last_heartbeat => 30.seconds.ago.utc) }
+    let(:stopping_worker) { FactoryBot.create(:miq_ui_worker, :status => MiqWorker::STATUS_STOPPING, :last_heartbeat => 30.seconds.ago.utc) }
+    let(:server)          { EvmSpecHelper.local_miq_server(:miq_workers => [started_worker, stopping_worker]) }
+
+    before do
+      allow(server).to receive(:workers_last_heartbeat) { Time.now.utc }
+      require 'miq-process'
+      allow(MiqProcess).to receive(:alive?).with(started_worker.pid).and_return(true)
+    end
+
+    it "validates current/starting workers for memory usage, avoiding stale miq_workers" do
+      server.sync_child_worker_settings
+      started_worker.update!(:unique_set_size => 4.gigabytes)
+
+      expect(server.post_message_for_workers(started_worker.class.name)).to eq([started_worker.id])
+      expect(MiqWorker.exists?(started_worker.id)).to be_falsy
+      expect(server.post_message_for_workers(started_worker.class.name)).to eq([])
+    end
+  end
```

Comment:
Does it make sense to have this test or is it too time dependent?

Reply:
I'm okay with this one. I don't think we're really relying on the timing too much here though, right? We're killing the worker because of memory, not a timeout.

Reply:
Do we want a similar spec on master?

Reply:
> I'm okay with this one. I don't think we're really relying on the timing too much here though, right? We're killing the worker because of memory, not a timeout

Right, the second call of the method needs to update the heartbeat (because it's newer) so as long as we don't run the test in microseconds, the new heartbeat should be newer and therefore cause the "update" that would normally trigger the can't modify frozen hash.

> Do we want a similar spec on master?

So, I tried to test this on master and was able to confirm it's not a problem there but because the restart/kill of the row and the update of the heartbeat happens from 2 separate methods on master, I had to comment out most of the monitor_workers code other than the loop to verify it wasn't affected.  So, maybe?